### PR TITLE
Add LCU and TRGMUX RTDs for S32K for Quadrature Decoding

### PR DESCRIPTION
**Pull request by sumitbatra-nxp**
We are adding -
1. Lcu and Trgmux RTDs for S32K
2. A counter to track the overflow interrupts on an eMIOS channel,
3.  S32DS generated headers for Lcu and Trgmux to support Quadrature decoder.

### CODE REVIEW

**Comment by manuargue** [comment]
+ @Dat-NguyenDuy  as reviewer

File: s32/drivers/s32k3/Mcl/src/Lcu_Ip.c [hunk 1]
```diff
@@ -0,0 +1,2260 @@
...
+*                                        GLOBAL VARIABLES
+==================================================================================================*/
+#define MCL_START_SEC_VAR_CLEARED_UNSPECIFIED_NO_CACHEABLE
+#include "Mcl_MemMap.h"
+
+/* Pointer to the LCU Initialization Configuration. Based on this configuration pointer,
+   the LCU Driver obtains all information for the Logic Cell and corresponding
+   Cell Configurations.
+   The Pointer is loaded when Lcu_Ip_Init() is called. */
+static const Lcu_Ip_InitType * Lcu_Ip_pxInit;
```
Comment (manuargue) [inline s32/drivers/s32k3/Mcl/src/Lcu_Ip.c:172]: all variables that are no-cacheable (ie. between `#define *START*NO_CACHEABLE` and `#define *STOP*NO_CACHEABLE`)  must be patched with `VAR_SEC_NOCACHE()` macro. Please do all patches in a separate commit after adding the driver sources. See https://github.com/zephyrproject-rtos/hal_nxp/commit/5c2af05d593698da333e56c564828814823dd081

**Review by manuargue — CHANGES_REQUESTED** [review]

File: s32/drivers/s32k3/Icu/include/Emios_Icu_Ip.h [hunk 2]
```diff
@@ -401,6 +401,23 @@ eMios_Icu_ValueType Emios_Icu_Ip_GetEdgeNumbers
...
+* @details    This function returns Overflow of MCB counter of eMIOS channel
+*
+* @param[in]  instance    - eMIOS module index
+* @param[in]  hwChannel   - eMIOS encoded hardware channel
+*
+* @return     uint32 - Counted Overflow instances
+*
+* @api
+*/
+eMios_Icu_ValueType Emios_Icu_Ip_GetOverflowCount
```
[the remark this answers is not in the document]
Comment (sumitbatra-nxp) [reply s32/drivers/s32k3/Icu/include/Emios_Icu_Ip.h]: Fixed in hal ! Will submit the corresponding zephyr patch in a short while

[the remark this answers is not in the document]
Comment (sumitbatra-nxp) [reply s32/drivers/s32k3/Icu/include/Emios_Icu_Ip.h]: Fixed in hal ! Will submit the corresponding zephyr patch in a short while

File: s32/drivers/s32k3/Mcl/src/Lcu_Ip.c (hunk 1 above)
Comment (sumitbatra-nxp) [reply s32/drivers/s32k3/Mcl/src/Lcu_Ip.c]: @manuargue I have discussed with Lcu and Trgmux IP owners and have resolved it.. kindly check !!

**Comment by manuargue** [comment]
@Dat-NguyenDuy kindly review this pr

File: s32/soc/s32k344/include/Emios_Icu_Ip_Defines.h [hunk 3]
```diff
@@ -150,13 +155,29 @@ extern "C"{
...
 #define EMIOS_ICU_IP_COUNTER_MASK              ((uint32)65535)
 #endif
+
+/* Driver writer needs to define a custom array of indices if,
+ *  Emios Icu channels are used for both PWM SAIC and MCB modes
+ *  as we cannot assume the number channels used
+ */
+#if (DT_HAS_COMPAT_STATUS_OKAY(nxp_qdec_s32) && !(DT_FOREACH_STATUS_OKAY(nxp_s32_emios_pwm, PWM_NXP_S32_CAPTURE_USED)))
+
+#define EMIOS_ICU_IP_INITIAL_INDEX_OF_CHANNELS \
```
Comment (Dat-NguyenDuy) [inline s32/soc/s32k344/include/Emios_Icu_Ip_Defines.h:165]: I don't think is a good idea to limit application by using pwm capture or qdec exclusively.
Beside that seems that you are hardcoding the qdec only uses emios channel 6 and 7,
which is not scalable

**Comment by Dat-NguyenDuy** [comment]
> @Dat-NguyenDuy kindly review this pr

Left some comments about what obviously i see on both PR. I will allocate time to take deeper into the PRs this week

File: s32/soc/s32k344/include/Emios_Icu_Ip_SA_Init_PBcfg.h [hunk 4]
```diff
@@ -1034,7 +1036,9 @@ extern "C"{
 #endif
 #endif /* DT_FOREACH_CHILD_STATUS_OKAY_VARGS */
 
-#define EMIOS_ICU_CONFIG_SA_INIT_PB
+#define EMIOS_ICU_CONFIG_SA_INIT_PB \
+  extern const eMios_Icu_Ip_ChannelConfigType eMios_Icu_Ip_0_ChannelConfig_PB_Init[2U]; \
```
[the remark this answers is not in the document]
Comment (sumitbatra-nxp) [reply s32/soc/s32k344/include/Emios_Icu_Ip_SA_Init_PBcfg.h]: fixed !

File: s32/soc/s32k344/include/Emios_Icu_Ip_Defines.h (hunk 3 above)
Comment (sumitbatra-nxp) [reply s32/soc/s32k344/include/Emios_Icu_Ip_Defines.h]: fixed as per your  suggestions

File: dts/nxp/s32/S32K344-172MQFP-pinctrl.h [hunk 5]
```diff
@@ -179,6 +179,7 @@
 #define PTA16_EMIOS_0_CH11_H_I          NXP_S32_PINMUX(0, 16, 0, 59, 2)
 #define PTA17_EMIOS_0_CH6_G_O           NXP_S32_PINMUX(0, 17, 2, 0, 0)
 #define PTA17_EMIOS_0_CH6_G_I           NXP_S32_PINMUX(0, 17, 0, 54, 2)
+#define TRGMUX_INT_OUT37_EMIOS_0_CH6_G  NXP_S32_PINMUX(0, 17, 0, 54, 3)
```
Comment (manuargue) [inline dts/nxp/s32/S32K344-172MQFP-pinctrl.h:182]: can you pls group them at the bottom of the file under `/* TRGMUX */`? this file was auto-generated with an internal script (yet to be published), let's keep the autogenerated part intact for easier maintenance. When the script is published it should also consider adding TRGMUX internal muxing for SIUL2.

**Review by manuargue — CHANGES_REQUESTED** [review]

File: dts/nxp/s32/S32K344-172MQFP-pinctrl.h (hunk 5 above)
Comment (sumitbatra-nxp) [reply dts/nxp/s32/S32K344-172MQFP-pinctrl.h]: Done !

**Comment by manuargue** [comment]
@sumitbatra-nxp are you planning to add all the missing pinmuxes also in the next updates? because right now only two combinations are added, and this quite limits its usage. For the rest of the changes looks fine with me.

@Dat-NguyenDuy  kindly review

**Comment by sumitbatra-nxp** [comment]
@manuargue .. I am not sure if we really want to handle these internal connections via pinmux driver in the longer run ?
As we thought of having an API in Siul to handle these cases directly

**Review by Dat-NguyenDuy — APPROVED** [review]

**Review by bperseghetti — APPROVED** [review]

**Review by manuargue — APPROVED** [review]
